iopath/common/azure_blob.py

```diff
@@ -0,0 +1,631 @@
+        Returns:
+            bool: true if the path is a file
+        """
+        self._check_kwargs(kwargs)
+
+        try:
+            props = self._get_blob_properties(path)
+            return props is not None
+        except azure_exceptions.AzureError as e:
+            logger.exception(e)
+            return False
+
+    def _isdir(self, path: str, **kwargs: Any) -> bool:
+        """
+        Checks if the resource at the given URI is a directory.
+        Args:
+            path (str): A URI supported by this PathHandler
+        Returns:
+            bool: true if the path is a directory
+        """
+        self._check_kwargs(kwargs)
+
+        _, _, dirpath = self._parse_uri(path)
+
+        try:
+            # Enumeration should find at least 1 longer child path
+            blob = next(self._enumerate_blobs(path))
+            return len(blob.name) > len(dirpath)
+        except StopIteration:
+            return False
+        except azure_exceptions.AzureError as e:
+            logger.exception(e)
+            return False
+
+    def _ls(self, path: str, **kwargs: Any) -> List[str]:
+        """
+        List the contents of the directory at the provided URI.
+        Args:
+            path (str): A URI supported by this PathHandler
+        Returns:
+            List[str]: list of contents in given path
+        """
+        self._check_kwargs(kwargs)
+
+        return [blob.name for blob in self._enumerate_blobs(path)]
+
+    def _local_cache_path(self, path: str):
+        """
+        Helper that returns a local cache path for a given uri.
+        Args:
+            path (str): A URI supported by this PathHandler.
+        Returns:
+            local_cache_path (str): a file path which exists on the local file system,
+            in a cache directory.
+        """
+        _, _, blob = self._parse_uri(path)
+        return os.path.join(get_cache_dir(self.cache_dir), self.CACHE_SUBDIR_NAME, blob)
+
+    def _get_local_path(self, path: str, force: bool = False, **kwargs: Any) -> str:
+        """
+        Get a filepath which is compatible with native Python I/O such as `open`
+        and `os.path`.
+        If URI points to a remote resource, this function may download and cache
+        the resource to local disk. In this case, the cache stays on filesystem
+        (under `file_io.get_cache_dir()`) and will be used by a different run.
+        Therefore this function is meant to be used with read-only resources.
+        Args:
+            path (str): A URI supported by this PathHandler
+        Returns:
+            local_path (str): a file path which exists on the local file system
+        """
+        self._check_kwargs(kwargs)
+
+        assert self._exists(path)
+        account, container, _ = self._parse_uri(path)
+        result_path = self._local_cache_path(path)
+
+        # TODO: this can be parallelized for directories
+        for blob in self._enumerate_blobs(path):
```

Comment:
A path should point to a single file (blob) in azure blob container. Why do you need to enumerate all the blobs here?

Reply:
This adds support for downloading both files and directories without any extra checks to see whether we're dealing with one case or the other. Azure's ListBlobs operation will return a singleton list when the "prefix" happens to match the the file path fully.